src/main/navigation/navigation.c

```diff
@@ -3087,6 +3090,35 @@ void updateFlightBehaviorModifiers(void)
     posControl.flags.isGCSAssistedNavigationEnabled = IS_RC_MODE_ACTIVE(BOXGCSNAV);
 }
 
+
+bool isOutsideGeofence(void)
+{
+    return (navConfig()->general.geofence_radius > 0 && GPS_distanceToHome > navConfig()->general.geofence_radius) || (navConfig()->general.geofence_height > 0 && posControl.actualState.abs.pos.z > navConfig()->general.geofence_height * 100);
+}
+
+bool isInsideBufferedGeofence(void)
+{
+    uint32_t bufferedRadius = navConfig()->general.geofence_radius - navConfig()->general.geofence_buffer;
+    uint32_t bufferedHeight = navConfig()->general.geofence_height - navConfig()->general.geofence_buffer;
+    return GPS_distanceToHome < (bufferedRadius > 0 ? bufferedRadius : 0) && posControl.actualState.abs.pos.z < (bufferedHeight > 0 ? bufferedHeight : 0) * 100;
+}
+
+void processGeofenceStatus(void)
+{
+    static bool geofenceTriggered = false;
+
+    if (!geofenceTriggered && isOutsideGeofence()) {
+        geofenceTriggered = true;
+        // No need to update status if F/S has already been enforced
+        if (getStateOfForcedRTH() == RTH_IDLE) {
+                activateForcedRTH();
```

Comment:
Which is the right way to trigger RTH and/or check if it's already active? `activateForcedRTH()`, `failsafeActivate(FAILSAFE_RETURN_TO_HOME)`, or both? The functionality of `activateForcedRTH()` looks controversial to me.